src/hypothesis/extra/django/models.py

```diff
@@ -17,159 +17,332 @@
-            dm.PositiveIntegerField: st.integers(0, 2147483647),
-            dm.PositiveSmallIntegerField: st.integers(0, 32767),
-            dm.BinaryField: st.binary(),
-            dm.BooleanField: st.booleans(),
-            dm.DateTimeField: datetimes(allow_naive=False),
-            dm.FloatField: st.floats(),
-            dm.NullBooleanField: st.one_of(st.none(), st.booleans()),
-        }
-    return __default_field_mappings
+_field_mappings = {
+    dm.AutoField: default_value,
+}
 
 
 def add_default_field_mapping(field_type, strategy):
-    field_mappings()[field_type] = strategy
-
-
-default_value = UniqueIdentifier(u'default_value')
+    _field_mappings[field_type] = strategy
 
 
-class UnmappedFieldError(Exception):
-    pass
-
-
-def validator_to_filter(f):
-    """Converts the field run_validators method to something suitable for use
-    in filter."""
-
+def validator_to_filter(field):
     def validate(value):
         try:
-            f.run_validators(value)
+            field.run_validators(value)
             return True
         except ValidationError:
             return False
-
     return validate
 
 
-def _get_strategy_for_field(f):
-    if isinstance(f, dm.AutoField):
-        return default_value
-    elif f.choices:
-        choices = [value for (value, name) in f.choices]
-        if isinstance(f, (dm.CharField, dm.TextField)) and f.blank:
-            choices.append(u'')
-        strategy = st.sampled_from(choices)
-    elif isinstance(f, dm.EmailField):
-        return ff.fake_factory(u'email')
-    elif type(f) in (dm.TextField, dm.CharField):
-        strategy = st.text(min_size=(None if f.blank else 1),
-                           max_size=f.max_length)
-    elif type(f) == dm.DecimalField:
-        m = 10 ** f.max_digits - 1
-        div = 10 ** f.decimal_places
-        q = Decimal('1.' + ('0' * f.decimal_places))
-        strategy = (
-            st.integers(min_value=-m, max_value=m)
+def field_strategy(*field_types, blank_choices=()):
+    """
+    Decorates the given field strategy factory to add support for:
+
+    - Passing in min_size and max_size kwargs for the field.
+    - Blank fields.
+    - Null fields.
+    - Filtering by field validators.
+
+    The field strategy is registered for the given fields.
+    """
+    def decorator(func):
+        @wraps(func)
+        def create_field_strategy(field, **kwargs):
+            # Handle field choices.
+            if field.choices:
+                choices = [
+                    value
+                    for (value, name)
+                    in field.choices
+                ]
+                if field.blank:
+                    choices.extend(blank_choices)
+                strategy = st.sampled_from(choices)
+            else:
+                # Run the factory function.
+                strategy = func(field, **kwargs)
+            # Add in null values.
+            if field.null:
+                strategy = st.one_of(strategy, st.none())
```

Comment:
Due to technical reasons of how shrinking works, this is better written `st.one_of(st.none(), strategy)` (shrinking will try to shrink towards the left of one_of)